app/controllers/application_controller.rb

```diff
@@ -5,6 +5,178 @@ class ApplicationController < Sinatra::Base
+  else
+  erb :"/users/login"
+end
+end
+
+get '/tweets' do
+  if logged_in?
+  @user = User.find(session[:user_id])
+ session[:user_id] = @user.id
+  erb :"/tweets/tweets"
+else
+  redirect '/users/login'
+end
+end
+
+ get '/tweets/new' do
+  #  raise session.inspect
+  # @user = User.find(session[:user_id])
+   if   logged_in?
+     erb :"/tweets/new"
+   else
+     redirect '/login'
+   end
+ end
+
+ get '/tweets/:id' do
+
+    @tweet = Tweet.find(params[:id])
+    # if logged_in?
+   if session[:user_id] == @tweet.user_id
+
+     @user = User.find(session[:user_id])
+   erb :"tweets/show"
+ elsif logged_in? && session[:user_id] != @tweet.user_id
+   @user = User.find(session[:user_id])
+   erb :"tweets/show"
+
+ else
+
+   redirect '/login'
+ end
+ end
+
+ get '/tweets/:id/edit' do
+   if logged_in?
+     @user = User.find(session[:user_id])
+     @tweet = Tweet.find(params[:id])
+     if @user.id == @tweet.user_id
+
+     erb :"/tweets/edit"
+   else
+     redirect '/tweets'
+   end
+ else
+   redirect '/login'
+ end
+ end
+
+ get "/users/:slug" do
+   @user = User.find_by_slug(params[:slug])
+     erb :'/users/show'
+ end
+
+get '/logout' do
+    session.clear
+  redirect '/login'
+ end
+
+patch '/tweets/:id' do
+  # binding.pry
+  if logged_in?
+   @tweet=Tweet.find(params[:id])
+     if !params[:content].empty?
+       @tweet.update(content: params[:content])
+
+      session[:tweet] = params[:content]
+      @user = User.find(session[:user_id])
+      @user.id = @tweet.user_id
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect "/tweets/#{@tweet.id}/edit"
+    end
+
+  end
+end
+
+post '/show' do
+  @user=User.find(session[:user_id])
+  @tweet = Tweet.new(content: params[:content])
+  if params[:content].empty?
+    redirect '/tweets/new'
+  else
+  @tweet.save
+  session[:tweet] = params[:content]
+
+  @user.tweets << @tweet
+
+  erb :"/users/show"
+end
+end
+
+post '/signup' do
+
+    if params[:username].empty? || params[:email].empty? || params[:password].empty?  #&& !logged_in?
+         redirect "/signup"
+     else
+       @user = User.create(:username => params[:username], :email => params[:email], :password => params[:password])
+       @user.save
+       if  logged_in? || @user.save
+       session[:user_id] = @user.id
+       session[:email] = @user.email
+       session[:username] = @user.username
+
+       redirect '/tweets'
+     end
+    end
+end
+
+
+post "/login" do
+     user = User.find_by(username: params[:username])
+    # if user
+    if user && user.authenticate(params[:password]) #|| logged_in?
+      # User.find_by(username: params[:username])
+          session[:user_id] = user.id
+          session[:email] = user.email
+          session[:username] = user.username
+        redirect "/tweets"
+    else
+        redirect "/login"
+    end
+end
+
+delete '/tweets/:id/delete' do
+
+    @tweet =Tweet.find_by_id(params[:id])
+   if  @tweet.user_id == session[:user_id] && logged_in?
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to. Prefer t